File: cursed_time/lib/src/balatro/menu.py
from enum import Enum
from typing import Optional
from dataclasses import dataclass

class Deck(Enum):
    """Deck types in Balatro"""
    Red = "b_red"
    Blue = "b_blue"
    Yellow = "b_yellow"
    Green = "b_green"
    Black = "b_black"
    Magic = "b_magic"
    Nebula = "b_nebula"
    Ghost = "b_ghost"
    Abandoned = "b_abandoned"
    Checkered = "b_checkered"
    Zodiac = "b_zodiac"
    Painted = "b_painted"
    Anaglyph = "b_anaglyph"
    Plasma = "b_plasma"
    Erratic = "b_erratic"

class Stake(Enum):
    """Stake levels in Balatro"""
    White = 1
    Red = 2
    Green = 3
    Black = 4
    Blue = 5
    Purple = 6
    Orange = 7
    Gold = 8

@dataclass
class Seed:
    """Game seed for reproducible runs"""
    value: str

class Menu:
    """Main menu screen handler"""
    def __init__(self, connection):
        self.connection = connection
    
    async def new_run(self, deck: Deck, stake: Stake, seed: Optional[Seed] = None):
        """Start a new run with specified parameters"""
        from .blinds import SelectBlind
        
        request_body = {
            "back": deck.value,
            "stake": stake.value,
            "seed": seed.value if seed else None
        }
        
        response = await self.connection.send_request("main_menu/start_run", request_body)
        
        if "Ok" in response["body"]:
            blind_info = response["body"]["Ok"]
            return SelectBlind(blind_info, self.connection)
        else:
            raise Exception(f"Failed to start run: {response['body']['Err']}")

File: client/lib/src/balatro/menu.rs
use crate::balatro::blinds::SelectBlind;
use crate::net::Connection;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use serde_repr::{Deserialize_repr, Serialize_repr};

pub struct Menu<'a> {
    connection: &'a mut Connection,
}

impl<'a> Menu<'a> {
    pub(crate) fn new(connection: &'a mut Connection) -> Self {
        Self { connection }
    }

    pub async fn new_run(
        self,
        deck: Deck,
        stake: Stake,
        seed: Option<Seed>,
    ) -> Result<SelectBlind<'a>, super::Error> {
        let new_run = protocol::StartRun {
            back: deck,
            stake,
            seed,
        };
        let blinds = self.connection.request(new_run).await??;
        Ok(SelectBlind::new(blinds, self.connection))
    }
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
pub enum Deck {
    #[serde(rename = "b_red")]
    Red,
    #[serde(rename = "b_blue")]
    Blue,
    #[serde(rename = "b_yellow")]
    Yellow,
    #[serde(rename = "b_green")]
    Green,
    #[serde(rename = "b_black")]
    Black,
    #[serde(rename = "b_magic")]
    Magic,
    #[serde(rename = "b_nebula")]
    Nebula,
    #[serde(rename = "b_ghost")]
    Ghost,
    #[serde(rename = "b_abandoned")]
    Abandoned,
    #[serde(rename = "b_checkered")]
    Checkered,
    #[serde(rename = "b_zodiac")]
    Zodiac,
    #[serde(rename = "b_painted")]
    Painted,
    #[serde(rename = "b_anaglyph")]
    Anaglyph,
    #[serde(rename = "b_plasma")]
    Plasma,
    #[serde(rename = "b_erratic")]
    Erratic,
}

#[derive(Serialize_repr, Deserialize_repr, Clone, Copy, Debug)]
#[repr(u8)]
pub enum Stake {
    White = 1,
    Red = 2,
    Green = 3,
    Black = 4,
    Blue = 5,
    Purple = 6,
    Orange = 7,
    Gold = 8,
}

impl FromStr for Deck {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.trim().to_lowercase().as_str() {
            "red" => Ok(Deck::Red),
            "blue" => Ok(Deck::Blue),
            "yellow" => Ok(Deck::Yellow),
            "green" => Ok(Deck::Green),
            "black" => Ok(Deck::Black),
            "magic" => Ok(Deck::Magic),
            "nebula" => Ok(Deck::Nebula),
            "ghost" => Ok(Deck::Ghost),
            "abandoned" => Ok(Deck::Abandoned),
            "checkered" => Ok(Deck::Checkered),
            "zodiac" => Ok(Deck::Zodiac),
            "painted" => Ok(Deck::Painted),
            "anaglyph" => Ok(Deck::Anaglyph),
            "plasma" => Ok(Deck::Plasma),
            "erratic" => Ok(Deck::Erratic),
            _ => Err(format!("Invalid deck. Valid options are: {}", 
                ["Red", "Blue", "Yellow", "Green", "Black", "Magic", "Nebula", 
                 "Ghost", "Abandoned", "Checkered", "Zodiac", "Painted", 
                 "Anaglyph", "Plasma", "Erratic"].join(", ")))
        }
    }
}

impl FromStr for Stake {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.trim().to_lowercase().as_str() {
            "white" => Ok(Stake::White),
            "red" => Ok(Stake::Red),
            "green" => Ok(Stake::Green),
            "black" => Ok(Stake::Black),
            "blue" => Ok(Stake::Blue),
            "purple" => Ok(Stake::Purple),
            "orange" => Ok(Stake::Orange),
            "gold" => Ok(Stake::Gold),
            _ => Err(format!("Invalid stake. Valid options are: {}",
                ["White", "Red", "Green", "Black", "Blue", "Purple", 
                 "Orange", "Gold"].join(", ")))
        }
    }
}

#[derive(Serialize)]
pub struct Seed(String);

pub(crate) mod protocol {
    use super::{Deck, Seed, Stake};
    use crate::{
        balatro::blinds::protocol::BlindInfo,
        net::protocol::{Packet, Request},
    };
    use serde::Serialize;

    // Hide serialization impls here since they're specific to Balatro's
    // internals.

    #[derive(Serialize)]
    pub struct StartRun {
        pub back: Deck,
        pub stake: Stake,
        pub seed: Option<Seed>,
    }

    impl Request for StartRun {
        type Expect = Result<BlindInfo, String>;
    }

    impl Packet for StartRun {
        fn kind() -> String {
            "main_menu/start_run".to_string()
        }
    }
}
